File: src/codes/Users_Manual/advanced/custom_modulation_0.cpp
//~#include "autd3.hpp"
//~#include <vector>
//~#include <limits>
class BurstModulation final
    : public autd3::derive::Modulation<BurstModulation> {
 public:
  std::vector<uint8_t> calc(const autd3::Geometry&) const override {
    std::vector buffer(_buf_size, std::numeric_limits<uint8_t>::min());
    buffer[_buf_size - 1] = std::numeric_limits<uint8_t>::max();
    return buffer;
  }

  explicit BurstModulation(const size_t buf_size = 4000) noexcept
      : Modulation(autd3::SamplingConfig::Freq(4000 * autd3::Hz)),
        _buf_size(buf_size) {}

 private:
  size_t _buf_size;
};
//~int main() { return 0; }

File: src/codes/Users_Manual/advanced/custom_modulation_0.rs
use autd3::prelude::*;
use autd3::derive::*;

#[derive(Modulation)]
pub struct Burst {
    config: SamplingConfig,
    loop_behavior: LoopBehavior,
}

impl Burst {
    pub fn new() -> Self {
        Self { 
            config: SamplingConfig::Freq(4000 * Hz),
            loop_behavior: LoopBehavior::infinite(),
        }
    }
}

impl Modulation for Burst {
    fn calc(&self, geometry: &Geometry) -> Result<Vec<u8>, AUTDInternalError> {
        Ok((0..4000)
            .map(|i| if i == 3999 { u8::MAX } else { u8::MIN })
            .collect())
    }
}
# fn main() { 
# }